modules/capability/capability_test.go

```diff
@@ -3,101 +3,119 @@ package capability_test
-	"github.com/cosmos/cosmos-sdk/types/module"
+	moduletestutil "github.com/cosmos/cosmos-sdk/types/module/testutil"
 	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
-	"github.com/stretchr/testify/suite"
 
-	"github.com/cosmos/ibc-go/v7/testing/simapp"
+	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
 
 	"github.com/cosmos/ibc-go/modules/capability"
 	"github.com/cosmos/ibc-go/modules/capability/keeper"
 	"github.com/cosmos/ibc-go/modules/capability/types"
 )
 
-const memStoreKey = "memory:mock"
+const mockMemStoreKey = "memory:mock"
 
 type CapabilityTestSuite struct {
 	suite.Suite
-	cdc    codec.Codec
-	ctx    sdk.Context
-	app    *simapp.SimApp
+
+	cdc codec.Codec
+	ctx sdk.Context
+
 	keeper *keeper.Keeper
-	module module.AppModule
+
+	storeKey        *storetypes.KVStoreKey
+	memStoreKey     *storetypes.MemoryStoreKey
+	mockMemStoreKey *storetypes.MemoryStoreKey
 }
 
 func (suite *CapabilityTestSuite) SetupTest() {
-	checkTx := false
-	app := simapp.Setup(suite.T(), checkTx)
-	cdc := app.AppCodec()
-
-	// create new keeper so we can define custom scoping before init and seal
-	keeper := keeper.NewKeeper(cdc, app.GetKey(types.StoreKey), app.GetMemKey(types.MemStoreKey))
-
-	suite.app = app
-	suite.ctx = app.BaseApp.NewContext(checkTx, tmproto.Header{Height: 1})
-	suite.keeper = keeper
-	suite.cdc = cdc
-	suite.module = capability.NewAppModule(cdc, *keeper, false)
+	encodingCfg := moduletestutil.MakeTestEncodingConfig(capability.AppModuleBasic{})
+	suite.cdc = encodingCfg.Codec
+
+	suite.storeKey = storetypes.NewKVStoreKey(types.StoreKey)
+	suite.memStoreKey = storetypes.NewMemoryStoreKey(types.MemStoreKey)
+	suite.mockMemStoreKey = storetypes.NewMemoryStoreKey(mockMemStoreKey)
+
+	suite.ctx = suite.NewTestContext()
+	suite.keeper = keeper.NewKeeper(suite.cdc, suite.storeKey, suite.memStoreKey)
+}
+
+func (suite *CapabilityTestSuite) NewTestContext() sdk.Context {
+	db := dbm.NewMemDB()
+	cms := store.NewCommitMultiStore(db, log.NewNopLogger(), metrics.NewNoOpMetrics())
+	cms.MountStoreWithDB(suite.storeKey, storetypes.StoreTypeIAVL, db)
+	cms.MountStoreWithDB(suite.memStoreKey, storetypes.StoreTypeMemory, db)
+	cms.MountStoreWithDB(suite.mockMemStoreKey, storetypes.StoreTypeMemory, db)
+
+	err := cms.LoadLatestVersion()
+	suite.Require().NoError(err)
+
+	return sdk.NewContext(cms, tmproto.Header{}, false, log.NewNopLogger())
 }
 
 // The following test case mocks a specific bug discovered in https://github.com/cosmos/cosmos-sdk/issues/9800
 // and ensures that the current code successfully fixes the issue.
+// This test emulates statesync by firstly populating persisted state by creating a new scoped keeper and capability.
+// In-memory storage is then discarded by creating a new capability keeper and app module using a mock memstore key.
+// BeginBlock is then called to populate the new in-memory store using the persisted state.
 func (suite *CapabilityTestSuite) TestInitializeMemStore() {
```

Comment:
I spent a good bit of time trying to parse this testcase. So I've added a bunch of in-line comments as well as extra info to the godoc above. This was initially added by @AdityaSripal in https://github.com/cosmos/cosmos-sdk/pull/9835 but has since seen a whole bunch of devs chopping into the code. 

Hopefully its now clear what its doing and is easier to maintain.

Reply:
Thank you for this!